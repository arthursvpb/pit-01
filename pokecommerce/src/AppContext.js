import { createContext } from "react";
import { parseJwt, tokenKey } from "./utils/util";

const AppContext = createContext();

const searchParams = new URLSearchParams(window.location.search);

const initialState = {
  pokemons: [],
  wishlist: [],
  cart: [],
  pagination: {
    currentPage: searchParams.get("page") ?? 1,
    search: '',
    limit: 9,
    total: 0,
    totalPages: 0,
  },
  loggedUser: parseJwt(localStorage.getItem(tokenKey)),
};

const reducer = (state, action) => {
  switch (action.type) {
    case "SET_LOGGED_USER": {
      return {
        ...state,
        loggedUser: action.payload,
      };
    }

    case "SET_POKEMON": {
      return {
        ...state,
        pokemons: action.payload,
      };
    }

    case "SET_WISHLIST": {
      return {
        ...state,
        wishlist: action.payload,
      };
    }

    case "SET_PAGINATION": {
      return {
        ...state,
        pagination: action.payload,
      };
    }

    default: {
      return state;
    }
  }
};

export { initialState, reducer };

export default AppContext;
