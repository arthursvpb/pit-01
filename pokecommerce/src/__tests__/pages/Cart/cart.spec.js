import { render, fireEvent } from "@testing-library/react";
import AppContext, { initialState } from "../../../AppContext";

import { pokemonCharizard, pokemonMelMetal } from "../../test.utils";

import Cart from "../../../pages/Cart";

const CartWrapper = ({ state = initialState, dispatch }) => (
  <AppContext.Provider value={[state, dispatch]}>
    <Cart />
  </AppContext.Provider>
);

describe("Cart", () => {
  it("renders", () => {
    const { asFragment, container } = render(<CartWrapper />);

    const emptyState = container.querySelector(".empty-state");

    expect(emptyState).toBeTruthy();
    expect(emptyState.textContent).toBe(`You don't have any purchased pokemon`);

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders with one purchased pokemon", () => {
    const state = {
      ...initialState,
      me: {
        purchasedPokemon: [pokemonMelMetal],
      },
    };

    const { container } = render(<CartWrapper state={state} />);

    const melMetalContainer = container.querySelector('div[title="Melmetal"]');

    expect(container.querySelectorAll(".pokemon-card")).toHaveLength(1);
    expect(melMetalContainer).toBeTruthy();
    expect(melMetalContainer.querySelector(".lexicon-icon-heart")).toBeTruthy();
    expect(
      melMetalContainer.querySelectorAll(".pokemon-types__pokemon-type")
    ).toHaveLength(1);
    expect(
      melMetalContainer.querySelector(".pokemon-types__pokemon-type span")
        .textContent
    ).toBe("Steel");
  });

  it("renders with more than one pokemon", () => {
    const state = {
      ...initialState,
      wishlist: [pokemonCharizard],
      me: {
        purchasedPokemon: [pokemonMelMetal, pokemonCharizard],
      },
    };

    const { container } = render(<CartWrapper state={state} />);

    const charizardContainer = container.querySelector(
      'div[title="Charizard"]'
    );

    expect(container.querySelectorAll(".pokemon-card")).toHaveLength(2);
    expect(charizardContainer).toBeTruthy();
    expect(
      charizardContainer.querySelector(".lexicon-icon-heart-full")
    ).toBeTruthy();

    const charizardFavoriteButton = charizardContainer.querySelector(
      ".favorite-pokemon"
    );

    fireEvent.click(charizardFavoriteButton);
  });
});
