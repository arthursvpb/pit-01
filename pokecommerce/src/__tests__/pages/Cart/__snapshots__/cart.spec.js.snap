// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart renders 1`] = `
<DocumentFragment>
  <div
    class="mt-4 container"
  >
    <h1>
      Purchased Pokemons
    </h1>
    <div>
      <div
        class="row"
      >
        <span
          class="empty-state"
        >
          You don't have any purchased pokemon
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
